File: bindings/node/perf/perf_stop_token.ts
// SPDX-License-Identifier: MPL-2.0

'use strict';

// Wire-level shutdown token shared by all Node perf patterns.
//
// PERF_SINGLE_TEST_POLICY § 1.4 / PERF_MULTI_TEST_POLICY § 1.3.1 mandate
// using a single sentinel message instead of `senderDone` flags + short
// polling cadences. The sender emits this token once at phase end; the
// receiver, waiting on `poller.wait(events, -1)`, drains in-flight messages and exits when
// it sees the token. This mirrors `k_stop_token` / `is_stop_token` in
// `bindings/cpp/perf/single/common/perf_single_common.hpp` and keeps
// the wire bytes identical across language bindings.

const STOP_TOKEN_STRING = '__zlink_perf_stop__';
const STOP_TOKEN_BYTES = Buffer.from(STOP_TOKEN_STRING);

function isStopToken(buf) {
  if (!buf) {
    return false;
  }
  if (buf.length !== STOP_TOKEN_BYTES.length) {
    return false;
  }
  if (Buffer.isBuffer(buf)) {
    return buf.equals(STOP_TOKEN_BYTES);
  }
  if (buf instanceof Uint8Array) {
    return Buffer.from(buf.buffer, buf.byteOffset, buf.byteLength).equals(STOP_TOKEN_BYTES);
  }
  return false;
}

function isStopTokenParts(parts) {
  if (!parts || parts.length !== 1) {
    return false;
  }
  const part = parts[0];
  if (part && typeof part.size === 'function' && part.size() !== STOP_TOKEN_BYTES.length) {
    return false;
  }
  if ((Buffer.isBuffer(part) || part instanceof Uint8Array)
      && part.length !== STOP_TOKEN_BYTES.length) {
    return false;
  }
  const data = typeof part?.data === 'function' ? part.data() : part;
  return isStopToken(data);
}

module.exports = {
  STOP_TOKEN_STRING,
  STOP_TOKEN_BYTES,
  isStopToken,
  isStopTokenParts,
};

export {
  STOP_TOKEN_STRING,
  STOP_TOKEN_BYTES,
  isStopToken,
  isStopTokenParts,
};

File: bindings/python/perf/perf_stop_token.py
"""Wire-level stop token shared by python perf single/multi benchmarks.

PERF_SINGLE_TEST_POLICY § 1.4 and PERF_MULTI_TEST_POLICY § 1.3.1 require
that sender / phase termination is signalled on the data wire by sending
this literal token once, instead of using out-of-band ``threading.Event``
flags combined with short polling cadences. Receivers wait with
``poller.wait(events, -1)`` (signal-driven, indefinite) and exit when an inbound
message matches ``STOP_TOKEN``.
"""

from __future__ import annotations


STOP_TOKEN: bytes = b"__zlink_perf_stop__"


def is_stop_token(data) -> bool:
    """Return ``True`` if the raw bytes represent the wire stop token."""

    if data is None:
        return False
    if isinstance(data, (bytes, bytearray, memoryview)):
        return bytes(data) == STOP_TOKEN
    return False


def is_stop_token_in_parts(parts) -> bool:
    """Return ``True`` if any frame in ``parts`` matches the stop token.

    Most patterns send the token as a single-frame message, so checking
    the first frame is enough; we still scan all frames defensively for
    multi-frame transports such as router/dealer-router.
    """

    if not parts:
        return False
    for part in parts:
        if is_stop_token(part):
            return True
    return False
